wvd-templates/DSC/Functions.ps1

```
@@ -327,4 +337,155 @@ function IsRDAgentRegistryValidForRegistration {
+
+.PARAMETER AgentInstallerFolder
+Required path to MSI installer file
+
+.PARAMETER AgentBootServiceInstallerFolder
+Required path to MSI installer file
+#>
+function InstallRDAgents {
+    Param(
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$AgentInstallerFolder,
+        
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$AgentBootServiceInstallerFolder,
+    
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [string]$RegistrationToken,
+    
+        [Parameter(mandatory = $false)]
+        [switch]$EnableVerboseMsiLogging
+    )
+
+    $ErrorActionPreference = "Stop"
+
+    Write-Log -Message "Boot loader folder is $AgentBootServiceInstallerFolder"
+    $AgentBootServiceInstaller = LocateFile -SearchPath $AgentBootServiceInstallerFolder -Name "*.msi"
+
+    Write-Log -Message "Agent folder is $AgentInstallerFolder"
+    $AgentInstaller = LocateFile -SearchPath $AgentInstallerFolder -Name "*.msi"
+
+    if (!$RegistrationToken) {
+        throw "No registration token specified"
+    }
+
+    RunMsiWithRetry -programDisplayName "RDAgentBootLoader" -isUninstall -argumentList @("/x {A38EE409-424D-4A0D-B5B6-5D66F20F62A5}", "/quiet", "/qn", "/norestart", "/passive") -msiOutputLogPath "C:\Users\AgentBootLoaderUnInstall.txt" -msiLogVerboseOutput:$EnableVerboseMsiLogging
+
+    while ($true) {
+        try {
+            $oldAgent = Get-Package -ProviderName msi -Name "Remote Desktop Services Infrastructure Agent"
+        }
+        catch {
+            #Ignore the error if it was due to no packages being found.
+            if ($PSItem.FullyQualifiedErrorId -eq "NoMatchFound,Microsoft.PowerShell.PackageManagement.Cmdlets.GetPackage") {
+                break
+            }
+
+            throw;

```

I'm a little concerned about this. If there is some other error besides not found error with `Get-Package` command, this will fail the DSC. Instead, what do you think about logging a warning but still continue to install the agent as best efforts ?

---

I think it is better to throw an error so that the user is aware of the problem rather than putting them in a possibly unsupported state by trying to continue. The only error I can think of would be not having permissions to view the installed packages but I think the DSC runs as the SYSTEM user so that shouldn't be a problem